vm_eval.c

```diff
@@ -2169,35 +2179,55 @@ specific_eval(int argc, const VALUE *argv, VALUE self, int singleton, int kw_spl
 
 /*
  *  call-seq:
- *     obj.instance_eval(string [, filename [, lineno]] )   -> obj
- *     obj.instance_eval {|obj| block }                     -> obj
+ *    instance_eval {|self| ... } -> object
+ *    instance_eval(string, filename = nil, lineno = nil) -> string
+ *
+ *  With a block given, passes +self+ to the block
+ *  and returns the block's return value;
+ *  this gives the block access to instance variables and private methods in +self+:
```

Comment:
I don't think "access to instance variables and private methods" is a completely accurate description. For example, you can also define methods for the object:

```ruby
o = Object.new
o.instance_eval do
  def hello = puts "hello world"
end
o.hello # => "hello world"
```

I think the original description is more accurate:

> Evaluates a string containing Ruby source code, or the given block, within the context of the receiver.